be1-go/channel/consensus/mod.go

```diff
@@ -189,84 +239,552 @@ func (c *Channel) VerifyPublishMessage(publish method.Publish) error {
+		c.createConsensusInstance(data.InstanceID)
+	}
+
+	c.createMessageInstance(msg.MessageID, sender)
+
 	return nil
 }
 
 // ProcessConsensusElectAccept processes an elect accept action.
-func (c *Channel) processConsensusElectAccept(data messagedata.ConsensusElectAccept) error {
+func (c *Channel) processConsensusElectAccept(sender kyber.Point, msg message.Message) error {
+
+	var data messagedata.ConsensusElectAccept
 
-	err := data.Verify()
+	err := msg.UnmarshalData(&data)
 	if err != nil {
-		return xerrors.Errorf("invalid consensus#elect_accept message: %v", err)
+		return xerrors.Errorf("failed to unmarshal consensus#elect-accept: %v", err)
+	}
+
+	c.log.Info().Msg("received a consensus#elect-accept message")
+
+	err = data.Verify()
+	if err != nil {
+		return xerrors.Errorf("invalid consensus#elect-accept message: %v", err)
 	}
 
 	// check wether a message with the correct ID was received previously
 	_, valid := c.inbox.GetMessage(data.MessageID)
 
 	if !valid {
-		return xerrors.Errorf("message doesn't correspond to any received message")
+		return xerrors.Errorf(messageNotReceived)
+	}
+
+	messageState, ok := c.messageStates[data.MessageID]
+	if !ok {
+		return xerrors.Errorf(messageStateInexistant, data.MessageID)
+	}
+	messageState.Lock()
+	defer messageState.Unlock()
+	if data.Accept {
+		messageState.electAcceptNumber += 1
+	}
+
+	// Once all Elect_Accept have been received, proposer creates new prepare
+	// message
+	if messageState.electAcceptNumber < c.hub.GetServerNumber() {
+		return nil
+	}
+	if messageState.currentPhase != ElectAcceptPhase ||
+		!messageState.proposer.Equal(c.hub.GetPubKeyOrg()) {
+		return nil
+	}
+
+	consensusInstance, ok := c.consensusInstances[data.InstanceID]
+	if !ok {
+		return xerrors.Errorf(consensusInstanceInexistant, data.InstanceID)
+	}
+	consensusInstance.Lock()
+	defer consensusInstance.Unlock()
+
+	if consensusInstance.proposed_try >= consensusInstance.promised_try {
+		consensusInstance.proposed_try += 1
+	} else {
+		consensusInstance.proposed_try = consensusInstance.promised_try + 1
+	}
+
+	// For now the consensus always accept a true if it complete
+	consensusInstance.proposed_value = true
+
+	newData := messagedata.ConsensusPrepare{
+		Object:     "consensus",
+		Action:     "prepare",
+		InstanceID: data.InstanceID,
+		MessageID:  data.MessageID,
+
+		CreatedAt: time.Now().Unix(),
+
+		Value: messagedata.ValuePrepare{
+			ProposedTry: consensusInstance.proposed_try,
+		},
+	}
+
+	byteMsg, err := json.Marshal(newData)
+	if err != nil {
+		return xerrors.Errorf("failed to marshal new consensus#prepare message: %v", err)
+	}
+
+	err = c.publishNewMessage(byteMsg)
+	if err != nil {
+		return xerrors.Errorf("failed to send new consensus#prepare message: %v", err)
+	}
+
+	return nil
+}
+
+// ProcessConsensusPrepare processes a prepare action.
+func (c *Channel) processConsensusPrepare(msg message.Message) error {
+
+	var data messagedata.ConsensusPrepare
+
+	err := msg.UnmarshalData(&msg)
```

Comment:
This typo was causing the problem discussed on Slack.
```suggestion
	err := msg.UnmarshalData(&data)
```